File: LeetCode/Easy/2099-Find_Subsequence_Of_Length_K_With_The_Largest_Sum/2099-FindSubsequenceOfLengthKWithTheLargestSum.py
class Solution:
    def maxSubsequence(self, nums: list[int], k: int) -> list[int]:
        idx = [(num, i) for i, num in enumerate(nums)]

        greatest = sorted(
            idx,
            reverse=True,
            key=lambda x: x[0],
        )[:k]

        idx_greatest = sorted(greatest, key=lambda x: x[1])

        ans = [num for num, _ in idx_greatest]

        return ans

    def maxSubsequence1(self, nums: list[int], k: int) -> list[int]:
        aux = nums.copy()
        aux.sort(reverse=True)
        aux = aux[:k]

        n = len(nums)

        picked = [False] * n

        ans = []

        for _ in range(k):
            i = 0
            while i < n:
                if nums[i] in aux and not picked[i]:
                    break
                i += 1

            aux.remove(nums[i])
            picked[i] = True
            ans.append(nums[i])

        return ans


a = Solution()
print(a.maxSubsequence(nums=[2, 1, 3, 3], k=2))
print(a.maxSubsequence(nums=[-1, -2, 3, 4], k=3))
print(a.maxSubsequence(nums=[3, 4, 3, 3], k=2))
print(a.maxSubsequence(nums=[1, 3, 4, 5, 2], k=2))

File: LeetCode/Easy/2099-Find_Subsequence_Of_Length_K_With_The_Largest_Sum/2099-FindSubsequenceOfLengthKWithTheLargestSum.ex
defmodule Solution do
  @spec max_subsequence(nums :: [integer], k :: integer) :: [integer]
  def max_subsequence(nums, k) do
    idx_nums = Enum.with_index(nums)

    greatest = idx_nums
      |> Enum.sort(fn {a, _idx_a}, {b, _idx_b} ->
        a > b
      end)
      |> Enum.take(k)

    idx_greatest = greatest
      |> Enum.sort(fn {_a, idx_a}, {_b, idx_b} ->
        idx_a < idx_b
      end)

    ans = Enum.map(idx_greatest, fn {num, _idx} -> num end)

    ans
  end
end
